### ISSUE TITLE
OSX shared lib build

### ISSUE THREAD

**Issue by zenna** [dreal/dreal3#141]

**Comment by iamed2** [comment]
+1

**Comment by soonhokong** [comment]
Generating `.dylib` files are done (which is easy). The hard part is to adjust their RPATHs. I'll give you update later.

**Comment by soonhokong** [comment]
I find a way to all-included `libdreal.dylib` file. Could you test it? It's available at https://dl.dropboxusercontent.com/u/10377018/libdreal.dylib. I forget how to test it in Julia. It would be great if someone reminds me of how to do that.

**Comment by soonhokong** [comment]
For the record:

```
$ clang++ -fpic -shared -Wl,-all_load libdreal.a libcapd.a libibex.a libprim.a libClp.a libClpSolver.a libCoinUtils.a -Wl,-noall_load -o libdreal.dylib

$ otool -L libdreal.dylib
libdreal.dylib:
	libdreal.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 120.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1213.0.0)
```

**Comment by iamed2** [comment]
I was able to test the Dropbox dylib, and it worked! Lovely, thanks :)

Here's what I had to change to get DReal.jl working on recent Julia master on OS X 10.10 (for some reason the enum didn't work so I had to change that as well):

```diff
diff --git a/src/DReal.jl b/src/DReal.jl
index 3989b32..997a008 100644
--- a/src/DReal.jl
+++ b/src/DReal.jl
@@ -22,12 +22,12 @@ bin_dir = joinpath(prefix,"bin")
 lib_dir = joinpath(prefix,"lib")

 try
-  @compat Libdl.dlopen(joinpath(lib_dir, "libprim.so"), Libdl.RTLD_LAZY|Libdl.RTLD_DEEPBIND|Libdl.RTLD_GLOBAL)
-  @compat Libdl.dlopen(joinpath(lib_dir, "libCoinUtils.so"), Libdl.RTLD_LAZY|Libdl.RTLD_DEEPBIND|Libdl.RTLD_GLOBAL)
-  @compat Libdl.dlopen(joinpath(lib_dir, "libClp.so"), Libdl.RTLD_LAZY|Libdl.RTLD_DEEPBIND|Libdl.RTLD_GLOBAL)
-  @compat Libdl.dlopen(joinpath(lib_dir, "libcapd.so"), Libdl.RTLD_LAZY|Libdl.RTLD_DEEPBIND|Libdl.RTLD_GLOBAL)
-  @compat Libdl.dlopen(joinpath(lib_dir, "libibex.so"), Libdl.RTLD_LAZY|Libdl.RTLD_DEEPBIND|Libdl.RTLD_GLOBAL)
-  @compat Libdl.dlopen(joinpath(lib_dir, "libdreal.so"), Libdl.RTLD_LAZY|Libdl.RTLD_DEEPBIND|Libdl.RTLD_GLOBAL)
+  # @compat Libdl.dlopen(joinpath(lib_dir, "libprim.so"), Libdl.RTLD_LAZY|Libdl.RTLD_DEEPBIND|Libdl.RTLD_GLOBAL)
+  # @compat Libdl.dlopen(joinpath(lib_dir, "libCoinUtils.so"), Libdl.RTLD_LAZY|Libdl.RTLD_DEEPBIND|Libdl.RTLD_GLOBAL)
+  # @compat Libdl.dlopen(joinpath(lib_dir, "libClp.so"), Libdl.RTLD_LAZY|Libdl.RTLD_DEEPBIND|Libdl.RTLD_GLOBAL)
+  # @compat Libdl.dlopen(joinpath(lib_dir, "libcapd.so"), Libdl.RTLD_LAZY|Libdl.RTLD_DEEPBIND|Libdl.RTLD_GLOBAL)
+  # @compat Libdl.dlopen(joinpath(lib_dir, "libibex.so"), Libdl.RTLD_LAZY|Libdl.RTLD_DEEPBIND|Libdl.RTLD_GLOBAL)
+  @compat Libdl.dlopen(joinpath(lib_dir, "libdreal.dylib"), Libdl.RTLD_LAZY|Libdl.RTLD_DEEPBIND|Libdl.RTLD_GLOBAL)
 catch e
   println("Could not load required shared libraries")
   rethrow(e)
@@ -72,11 +72,11 @@ const DEFAULT_PRECISION = 0.001

 include("wrap_capi.jl")
 # Julia 0.3 does not have enums
-if VERSION >= v"0.4.0-dev"
-  include("logic.jl")
-else
+# if VERSION >= v"0.4.0-dev"
+#   include("logic.jl")
+# else
   include("logicv3.jl")
-end
+# end
 include("context.jl")
 include("environment.jl")
 include("expression.jl")
```

All you have to do to test is download the dylib to `~/.julia/v0.4/DReal/deps/usr/lib/libdreal.dylib`, open Julia and run `Pkg.test("DReal")`

**Comment by soonhokong** [comment]
@iamed2, thanks for checking things up! It's great to know that it's working!

 - I'll automate the process of generating `libdreal.dylib` on OSX so that you can make it by simply running `make shared_lib`.

 - For OSX, `libdreal.dylib` should be the only file that you need. That is, you don't really need to dlopen other `.so`. @iamed2, Could you check that?

 - @zenna, you taught me how to update dreal/dreal.jl repo. My understanding is that we need to update https://github.com/dreal/DReal.jl/blob/master/deps/build.jl file, right? In addition to the `dReal-$version-linux-shared-libs.tar.gz` file under the release, I'll add another one following a new nameing convention, `dReal-$version-darwin-shared-libs.tar.gz`. Could you update the build.jl file accordingly so that it will pick up the right file based on the platform?

**Comment by iamed2** [comment]
@soonhokong Yup, if you check the diff you'll see the .so loads are all commented out and it all still works :) (in fact, it doesn't work if they aren't commented out)

I can make a pull request for the relevant changes to the DReal.jl repo

**Comment by soonhokong** [comment]
@iamed2, right! Thanks again!!

**Comment by zenna** [comment]
Yes its build.jl.  We'll need to write some os specific code.  So for osx we only need to import 

```
+  @compat Libdl.dlopen(joinpath(lib_dir, "libdreal.dylib"), Libdl.RTLD_LAZY|Libdl.RTLD_DEEPBIND|Libdl.RTLD_GLOBAL)
```

?

None of the others are required?  I don't have a mac.

**Comment by soonhokong** [comment]
> None of the others are required?

Yes, that's all you need on a Mac.

> I don't have a mac.

Well.. I can test it. Just let me know when you need a test.

**Comment by zenna** [comment]
Try now!

**Comment by zenna** [comment]
Actually, you're better off taking that pull request..

**Comment by soonhokong** [comment]
@zenna, it says there are conflicts between what you just pushed and what @iamed2 added. Is it fine to take @iamed2's commit over yours?

**Comment by soonhokong** [comment]
I'm making `dReal-$version-darwin-shared-libs.tar.gz` file. When I get one, I'll add it to the release page, and test it on my mac.

**Comment by soonhokong** [comment]
Released: https://github.com/dreal/dreal3/releases/tag/v3.15.07.02

**Comment by soonhokong** [comment]
Closed by c346b86.

**Comment by soonhokong** [comment]
@zenna, I think it's not ready to test it yet, right? Let's continue the discussion at https://github.com/dreal/DReal.jl/issues/15.
